libafl/src/stages/pruning.rs

```diff
@@ -0,0 +1,151 @@
+//! Corpus pruning stage
+
+use core::marker::PhantomData;
+
+use libafl_bolts::{rands::Rand, Error};
+
+#[cfg(feature = "std")]
+use crate::{events::EventRestarter, state::Stoppable};
+use crate::{
+    corpus::Corpus,
+    stages::Stage,
+    state::{HasCorpus, HasRand, UsesState},
+};
+
+#[derive(Debug)]
+/// The stage to probablistically disable a corpus entry.
+/// This stage should be wrapped in a if stage and run only when the fuzzer perform restarting
+/// The idea comes from `https://mschloegel.me/paper/schiller2023fuzzerrestarts.pdf`
+pub struct CorpusPruning<EM> {
+    /// The chance of retaining this corpus
+    prob: f64,
+    phantom: PhantomData<EM>,
+}
+
+impl<EM> CorpusPruning<EM> {
+    fn new(prob: f64) -> Self {
+        Self {
+            prob,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<EM> Default for CorpusPruning<EM> {
+    fn default() -> Self {
+        Self::new(0.05)
+    }
+}
+
+impl<EM> UsesState for CorpusPruning<EM>
+where
+    EM: UsesState,
+{
+    type State = EM::State;
+}
+
+impl<E, EM, Z> Stage<E, EM, Z> for CorpusPruning<EM>
+where
+    EM: UsesState,
+    E: UsesState<State = Self::State>,
+    Z: UsesState<State = Self::State>,
+    Self::State: HasCorpus + HasRand,
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn perform(
+        &mut self,
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        state: &mut Self::State,
+        _manager: &mut EM,
+    ) -> Result<(), Error> {
+        // Iterate over every corpus entry
+        let n_corpus = state.corpus().count_all();
+        let mut do_retain = vec![];
+        for _ in 0..n_corpus {
+            let r = state.rand_mut().below(100) as f64;
+            do_retain.push((self.prob * 100_f64) < r);
+        }
+
+        for (i_th, retain) in do_retain.iter().enumerate().take(n_corpus) {
+            if !retain {
+                let corpus_id = state.corpus().nth_from_all(i_th);
+
+                let corpus = state.corpus_mut();
+                let removed = corpus.remove(corpus_id)?;
+                corpus.add_disabled(removed)?;
```

Comment:
Is there no way to set the state to disabled? Calling remove and re-adding sounds worse

Reply:
Looks like we don't have a method to disable, cc @R9295

Reply:
i don't think there's other way

Reply:
There isn't a method but I think makes sense to add one

Reply:
essentially disabled corpus and enabled corpus are separate corpus so there's no other way than removing and adding right?